### PR TITLE
[NTUSER] Increase desktop heap size

### PR DESCRIPTION
**Pull request by tkreuzer**
Distinguish between the following 3 scenarios:
1. Interactive Winsta0, "Winlogon" desktop: use 128 KB
2. Interactive Winsta0, "Default" & other desktop: use 3 MB (x86) / 20 MB (x64)
3. Non-interactive winsta: use 128 KB

This is what Windows 2003 uses for the interactive desktops on x86 and what Windows 7 uses for x64. Fixes desktop heap exhaustion during testing. The previous value was 512 KB for all desktops. It doesn't handle the even smaller "Disconnected" desktop, which is only meaningful for RDP sessions and we don't use it anyway. See comments in the file for references.

# Tests
KVM x86:  https://reactos.org/testman/compare.php?ids=97553,97557,97586,97589
- Fixes a lot of tests, no more additional reboots
- Seems to cause a crash in user32:menu (investigating)
- New run: https://eu3.proxysite.com/process.php?d=GcdT%2Fcwyyl4Byf8X0eJrE%2B5Pn5kN1ykUXvjMEeNt0YxPIySkqfdyJjqT3V9FGUNuEgrq%2BA0lmYD6&b=1
- Good to go.

KVM x64: https://reactos.org/testman/compare.php?ids=97559,97573,97584,97587,97590
- Improvement?

### CODE REVIEW

**Review by binarymaster — APPROVED** [review]

**Review by whindsaks — APPROVED** [review]

**Review by TAN-Gaming — APPROVED** [review]

**Review by DarkFire01 — APPROVED** [review]

**Review by Doug-Lyons — APPROVED** [review]

File: win32ss/user/ntuser/desktop.c [hunk 1]
```diff
@@ -2259,15 +2276,34 @@ IntPaintDesktop(HDC hDC)
...
+        else
+        {
+            HeapSize = gdwDesktopSectionSize * 1024;
+        }
+    }
+    else
+    {
+        HeapSize = gdwNOIOSectionSize * 1024;
+    }
+
```
Comment (SergeGautherie) [inline win32ss/user/ntuser/desktop.c:2306]: You might tweak this block with:
```suggestion

    /* Convert to bytes */
    HeapSize *= 1024;
```

File: win32ss/user/ntuser/desktop.c [hunk 2]
```diff
@@ -27,9 +27,26 @@ IntFreeDesktopHeap(IN PDESKTOP pdesk);
...
+ * - https://www.betaarchive.com/wiki/index.php?title=Microsoft_KB_Archive/184802
+ * - https://kb.firedaemon.com/support/solutions/articles/4000086192-windows-service-quota-limits-and-desktop-heap-exhaustion
+ * - https://learn.microsoft.com/en-us/troubleshoot/windows-server/performance/desktop-heap-limitation-out-of-memory
+ */
+#ifdef _WIN64
+DWORD gdwDesktopSectionSize = 20 * 1024; // 20 MB (Windows 7 style)
+#else
+DWORD gdwDesktopSectionSize = 3 * 1024; // 3 MB (Windows 2003 style)
+#endif
 DWORD gdwNOIOSectionSize    = 128; // A guess, for one or more of the first three system desktops.
```
Comment (SergeGautherie) [inline win32ss/user/ntuser/desktop.c:48]: Is the existing comment still relevant?
Why not `512` as in new detailed comment?

**Review by learn-more — APPROVED** [review]

File: win32ss/user/ntuser/desktop.c [hunk 3]
```diff
@@ -27,9 +27,26 @@ IntFreeDesktopHeap(IN PDESKTOP pdesk);
...
+ * Windows 7 x86: 12 MB / 512 KB
+ * Windows 7 x64: 20 MB / 768 KB
+ * Windows 10 x64: 20 MB / 4 MB
+ * See:
+ * - https://dbmentors.blogspot.com/2011/09/desktop-heap-overview.html
+ * - https://www.ibm.com/support/pages/using-microsoft-desktop-heap-monitor-dheapmon-determine-desktop-heap-space-and-troubleshoot-filenet-p8-and-image-services-issues
+ * - https://www.betaarchive.com/wiki/index.php?title=Microsoft_KB_Archive/184802
+ * - https://kb.firedaemon.com/support/solutions/articles/4000086192-windows-service-quota-limits-and-desktop-heap-exhaustion
+ * - https://learn.microsoft.com/en-us/troubleshoot/windows-server/performance/desktop-heap-limitation-out-of-memory
+ */
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/desktop.c:42]: Can we still boot ReactOS to Desktop with just 49MB RAM assigned to the VM after this change? 0.4.14-release-122-g7d2cd22 can successfully boot to desktop **in bootcd and livecd** with just 49MB assigned.

image

(bootcd-mem-requirements are usually a bit higher than livecd's)

File: win32ss/user/ntuser/desktop.c (hunk 2 above)
Comment (HBelusca) [inline win32ss/user/ntuser/desktop.c:48]:
```suggestion
DWORD gdwNOIOSectionSize    = 128;
```

Comment (HBelusca) [reply win32ss/user/ntuser/desktop.c:48]: The comment indeed doesn't seem relevant anymore, it is explained in the detailed one above.
```suggestion
DWORD gdwNOIOSectionSize    = 128;
```

**Review by HBelusca — APPROVED** [review]

File: win32ss/user/ntuser/desktop.c (hunk 3 above)
Comment (DarkFire01) [reply win32ss/user/ntuser/desktop.c:42]: image

File: win32ss/user/ntuser/desktop.c (hunk 1 above)
Comment (tkreuzer) [reply win32ss/user/ntuser/desktop.c:2306]: No

File: win32ss/user/ntuser/desktop.c (hunk 3 above)
Comment (tkreuzer) [reply win32ss/user/ntuser/desktop.c:42]: This change shouldn't affect memory requirements. Feel free to test.

File: win32ss/user/ntuser/desktop.c (hunk 2 above)
Comment (tkreuzer) [reply win32ss/user/ntuser/desktop.c]: I removed the comment. We don't need anything larger for the non-interactive desktops now, so I didn't want to change it.

**Pull request merged**